[build] Add Makefile build script generator

salkinium opened the pull request:
Makefiles are quite simple to use and I wanted a build system that didn't require Python3 (at least for building, all our tooling requires it of course). It may also make it easier for people wary of SCons or CMake to use modm.

- [x] Compilation and linking of .elf file
- [x] Compilation of source files with *custom* flags
-  ~Merging of Makefile generated by`::cmake` (almost identical)~ CMake Makefile will go away soon.
- [x] Move all special tools from `:::scons` to `::build` and also implement in `:::make`
  - [x] unit test tool
  - [x] information tool
  - [x] bitmap tool
- [x] Add Makefile generator to examples
- [x] Test Makefile Tools
  - [x] Bitmap tool
  - [x] infomation tool: git + build
  - [x] unit test
  - [x] program, program-bossac, program-fuses, program-dfu
  - [x] debug, debug-coredump with ui=tui and ui=web
  - [x] reset
- [x] Documentation

cc @rleh @chris-durand

rleh commented:
> - [ ]  Merging of Makefile generated by `::cmake` (almost identical)

I never understood why there is a Makefile generated by the cmake generator. With the rework of the cmake generator in #568 it will probably disappear?!

rleh commented on tools/build_script_generator/make/resources/Makefile.in:
```diff
@@ -0,0 +1,84 @@
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+APP_SRC := $(wildcard *.cpp)
```
Also include `*.c++`?

And maybe even C files?

salkinium replied:
Yes, just need to figure out how to search for nested sources.

rleh reviewed (approved):
Looks nice and clean! 👍🏽 

(and I like Makefiles 😆)

ghost commented:
> > * [ ]   Merging of Makefile generated by `::cmake` (almost identical)
> 
> I never understood why there is a Makefile generated by the cmake generator. With the rework of the cmake generator in #568 it will probably disappear?!

I tried to remove it but the CI is using it to build all cmake examples.

salkinium commented:
The Makefile is just there for tooling support, ie programming etc. The CI uses this Python script to compile the examples:
https://github.com/modm-io/modm/blob/develop/tools/scripts/examples_compile.py#L58

rleh commented:
>> ...
>
> I tried to remove it but the CI is using it to build all cmake examples.

That should be easily fixable in [the script](https://github.com/modm-io/modm/blob/develop/tools/scripts/examples_compile.py#L58) @salkinium pointed to:

```diff
-commands.append("make cmake && make build")
+commands.append("mkdir build && cd build && cmake .. && make")
```

---

> The Makefile is just there for tooling support, ie programming etc.

The correct way, as far as I know, to add tooling support to CMake would be to use [`add_custom_target(...)`](https://cmake.org/cmake/help/latest/command/add_custom_target.html).

salkinium committed:
[build] Add Makefile build script generator

salkinium committed:
[examples] Fix undefined IRQ example

salkinium committed:
[board] Fix SAMD21 MINI board GPIO definitions

salkinium commented:
Ok, this is ready for review now. It's a copy of the SCons "API" with all features except `make artifact` and there's no XPCC integration.

salkinium committed:
[examples] Add Makefile generator to select examples

rleh commented on examples/avr/assert/project.xml:
```diff
@@ -7,5 +7,6 @@
     <module>modm:architecture:delay</module>
     <module>modm:build:scons</module>
     <module>modm:build:cmake</module>
+    <module>modm:build:make</module>
```
```suggestion
    <module>modm:build:**</module>
```
... and consistent in all examples?

rleh replied:
Hmm ok, won't work with `tools/scripts/examples_compile.py`.

salkinium replied:
No, that'll require refactorings again in the future for excluding build systems that do not support all features.

rleh commented on tools/build_script_generator/cmake/module.lb:
```diff
@@ -132,7 +132,7 @@ def post_build(env):
     if env["include_cmakelists"]:
         env.template("resources/CMakeLists.txt.in", "CMakeLists.txt")
-    if env["include_makefile"]:
+    if env["include_makefile"] and not env.has_module("::make"):
         env.template("resources/Makefile.in", "Makefile")
```
As indicated earlier, I don't see the point of generating a Makefile from CMake module.

(But this should be ok for now, we can discuss this in #568.)

rleh commented on tools/build_script_generator/make/module.lb:
```diff
@@ -0,0 +1,135 @@
+from os.path import join, relpath, isdir, exists
+
+def init(module):
+    module.name = ":build:make"
```
```suggestion
    module.name = ":build:makefile"
```
Maybe?

salkinium replied:
Nah, the official name is GNU Make: https://www.gnu.org/software/make/.
We also don't call it `:build:sconstruct` or `:build:cmakelist`.
The hamming distance from `:build:cmake` is a little low, but then again it's a mistake you'll definitely notice.

rleh commented on tools/build_script_generator/make/module.md:
```diff
@@ -0,0 +1,509 @@
+**If you need something special, write your own Makefile, maybe starting by
+modifying ours.**
+Remember to set `modm:build:make:include_makefile` to `False`, so that your
+custom `Makefile` does not get overwritten.
```
```suggestion
custom `Makefile` does not get overwritten by `lbuild build`.
```

rleh commented on tools/build_script_generator/make/module.md:
```diff
@@ -0,0 +1,509 @@
+
+## Makefile Targets
+
+This module generates these `.PHONY` targets.
```
```suggestion
This module generates these [`.PHONY` targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html).
```

rleh commented on tools/build_script_generator/make/module.md:
```diff
@@ -0,0 +1,509 @@
+        compiler is called with.
+- `profile=release`: Compile project with the release profile options (default).
+- `profile=debug`: Compile project with the debug profile options.
+
```
```suggestion
- `-j8`: Use 8x parallelism.

```

rleh commented on tools/build_script_generator/make/module.md:
````diff
@@ -0,0 +1,509 @@
+section][fuses] mechanism, they are automatically used by avrdude. For fuse
+values see the [AVR Fuse Calculator][fusecalc].
+
+```c
+#include <avr/io.h>
+FUSES =
+{
+    LFUSE_DEFAULT, // .low
+    HFUSE_DEFAULT, // .high
+    EFUSE_DEFAULT, // .extended
+};
+```
````
Does this really belong here?

salkinium replied:
Good point, I'm moving it to `modm:platform:core` and linking it here.

rleh commented on tools/build_script_generator/make/module.md:
````diff
@@ -0,0 +1,509 @@
+
+```
+$ make program-bmp port=/dev/cu.usbmodemDEADBEEF
+Remote debugging using /dev/cu.usbmodemDEADBEEF
````
Not 
```suggestion
$ make program-bmp port=/dev/tty.usbmodemDEADBEEF
Remote debugging using /dev/tty.usbmodemDEADBEEF
```
like it says three lines above?

rleh commented on tools/build_script_generator/make/module.md:
````diff
@@ -0,0 +1,509 @@
+```
+
+Launches OpenOCD in the background, then launches GDB in foreground with the
+correct executable with text-based or [web-based GDBGUI](gdbgui) UI. When GDB
````
```suggestion
correct executable with text-based or [web-based GDBGUI][gdbgui] UI. When GDB
```

rleh replied:
And the link is missing at end of this file.

rleh commented on tools/build_script_generator/scons/module.md:
```diff
@@ -597,14 +586,19 @@ header file and add these two defines to the command line CPP options:
-You can enable this by setting the `modm:build:scons:info.git` option.
+You can enable this by setting the `modm:build:info.git` option.
+
+!!! info "Added delay"
```
```suggestion
!!! info "Increased build time"
```
I wouldn't call that a delay....

rleh commented on tools/scripts/examples_compile.py:
```diff
@@ -69,16 +68,24 @@ def build(project):
+		commands.append( ("scons run", "SCons") )
+	if ":build:make" in project_cfg:
+		commands.append( ("make run", "Make") )
+
```
What about cmake?

salkinium replied:
Eh, let's fix it in #568. I'd have to add a `make run` to the Makefile that will be removed…

salkinium commented:
Ah, of course, using shell commands doesn't work on Windows. That's going to be annoying…
```
Compiling C++Â·Â· release/modmsrcmodmdebugloggerhosteddefault_style.omkdir: unknown option -- M
74
Try 'mkdir --help' for more information.
```

salkinium commented:
Ok, so now that I have a Windows VM I can say that generating a Windows-compatible Makefile requires dealing **manually** (!!!) with the path seperator `\` vs `/` incl. in generated paths (ie. make does not care either). I've therefore decided not to care about Windows support for the Makefile generator (ie. you can use SCons or CMake) and will simply put a big warning about not using it on Windows.

salkinium commented:
This is ready for merging, unless someone™ wants to review.

salkinium committed:
[build] Move unittest.source to :build module

salkinium committed:
[build] Move info tools to :build module

salkinium commented:
I added an alias mechanism to lbuild v1.16.0 so that the move of `:build:scons:info.git` to `:build:info.git` is backwards compatible and prints a warning about what happened.


```
 $ lbuild -D :build:scons:info.git=Info discover -t :build:scons --developer
[WARNING] lbuild.node: Node 'modm:build:scons:info.git' has been moved to ':build:info.git'!

>> modm:build:scons:info.git  [Alias]

Moved into `modm:build` module

Alias: :build:info.git

>> modm:build:info.git  [EnumerationOption]

# Generate git repository state information

- `Info`: generates information about the last commit.
- `Info+Status`: like `Info` plus git file status.

Value: Disabled
Inputs: [Disabled, Info, Info+Status]

Module(modm:build:scons)   SCons Build Script Generator
├── Option(cache_dir) = [] in [Path]   Path to SConstruct CacheDir
├── Option(include_sconstruct) = True in [True, False]   Generate a SConstruct file
├── Option(path.artifact) = artifacts in [Path]   Path to Artifact Store
├── Collector(flag_format) in [Callable]   Formatting compile flags for SCons
├── Collector(path.tools) in [Path]   SCons tool paths to be added to the Environment
├── Collector(tools) in [String]   SCons tools to be added to the Environment
├── Alias(image.source) -> :build:image.source   Moved into `modm:build` module
├── Alias(info.build) -> :build:info.build   Moved into `modm:build` module
├── Alias(info.git) -> :build:info.git   Moved into `modm:build` module
╰── Alias(unittest.source) -> :build:unittest.source   Moved into `modm:build` module
```

salkinium commented:
Ping @rleh, please review so I can merge…

rleh reviewed (changes requested):
Make uses the build folder `make-release/`, so scons should also use `scons-release/` (currently `release/`).

rleh commented:
With the _stm32f469_discovery/hard_fault_ example the `*.elf` files are different, the one produced by Make is 228 Bytes smaller.

rleh commented on examples/nucleo_f103rb/undefined_irq/main.cpp:
```diff
@@ -71,7 +68,7 @@ int main()
 			NVIC_SetPendingIRQ(IRQn_Type(int(EXTI0_IRQn) + ii));
 			ii = (ii + 1) % 5;
-			// wait one second for debounce
+			// wait one time for debounce
```
What?

salkinium commented:
The interesting question is if the .bin file is different? I would assume so perhaps because of link order…

rleh commented on src/modm/board/samd21_mini/board.hpp:
```diff
@@ -86,10 +86,10 @@ using D12 = GpioA19;
-using LedTx = GpioInverted<GpioOutputA27>;
-using LedRx = GpioInverted<GpioOutputB03>;
-using Leds = SoftwareGpioPort< LedTx, LedRx >;
+using LedD13 = GpioA17;
+using LedTx = GpioA27;
+using LedRx = GpioB03;
+using Leds = LedTx;
```
```suggestion
// User LED (Rx and Tx LEDs inverted, because connected to 3V3)
using LedD13 = GpioA17;
using LedTx = GpioInverted<GpioOutputA27>;
using LedRx = GpioInverted<GpioOutputB03>;
using Leds = SoftwareGpioPort< LedTx, LedRx >;
```

![grafik](https://user-images.githubusercontent.com/2820734/112638306-66942200-8e3f-11eb-9b73-e3644afae47b.png)
https://cdn.sparkfun.com/datasheets/Dev/Arduino/Boards/sparkfun-samd21-mini-breakout-v10.pdf

salkinium replied:
I know, but the problem is that the SAM GPIO implementation does not have `GpioInverted` or `SoftwareGpioPort` implemented…

rleh replied:
Ohhh :D

Maybe we (you) should create a issue about that and add the "help wanted" label?

salkinium replied:
If I'm honest I would rather rework the [Targets support matrix](https://github.com/modm-io/modm#targets) to something that's generated from lbuild using the data we already accumulate in `generate_module_docs` (but perhaps not so slow). That would give a much more detailed look at what peripherals are supported.

salkinium replied:
Also not sure if the "help wanted" is generally useful, so far I've not seen a contributor specifically fix things they are not personally motivated to do.

rleh replied:
I think this is also because we hardly have such "help-wanted" problems that are suitable for beginners.
We could also try to participate in events like [hacktoberfest](https://hacktoberfest.digitalocean.com/) to attract new contributors.

rleh commented:
> The interesting question is if the .bin file is different? I would assume so perhaps because of link order…

Bin (and hex) files differ only slightly, probably because of GCCs build id?
![grafik](https://user-images.githubusercontent.com/2820734/112639795-f4244180-8e40-11eb-9a8e-c0ab63f46790.png)

I think that's fine!
